File: aufgabe3_cpp/delta/delta.cpp
#include "delta.hpp"

using std::string;
using std::map;
using std::vector;
using std::mutex;
using std::queue;
using std::pair;

namespace Delta
{
    //==========================================================================
    // RETURN CODES

//    const int EXIT_OK = 0;
//    const int EXIT_NO_INPUT_FILES = 1;
//    const int EXIT_ERROR_WORKDIR = 11;
//    const int EXIT_ERROR_OUTFILE = 12;

    //==========================================================================
    // CONFIGURATION

//    const string WORKDIR("./");
//    const string OUTFILE(WORKDIR + "horst.txt");
//    const int INPUTBUFFER_SIZE = 1000;
//    const int NUM_THREADS = 3;

    //==========================================================================
    // VARIABLES

    map<string, map<string, int> > RESULTS;
    vector<string> DONE_FILES;
    queue<pair<string, string> > JOBQUEUE;
    int DELTA_COUNT;
    mutex RESULT_LOCK;
    mutex JOBQUEUE_LOCK;

    //==========================================================================
    // FUNCTIONS

//    vector<string> getCsvFilesInDir(const string& path);
//    bool resultsExist(const string& f1, const string& f2);
//    string resultKey(const string& f1, const string& f2);
//    void addResultsForEquals(const string& f1, const string& f2);
//    void addResults(const string& f1, const string& f2, const map<string, int>& result);
//    void addFromExistingResults(const string& f1, const string& f2, const string& key);
//    string printResult(const string& f1, const string& f2, map<string, int>& result);
//    void startCalculator();
}

File: aufgabe3_ruby/delta/delta.rb
require "thread"
require_relative "calculation/entry"
require_relative "calculation/WindowDeltaCalculator"

class Delta
    EXIT_OK = 0
    EXIT_NO_INPUT_FILES = 1
    EXIT_ERROR_WORKDIR = 11
    EXIT_ERROR_OUTFILE = 12
    
    WORKDIR = "./"
    OUTFILE = WORKDIR + "horst.txt"
    INPUTBUFFER_SIZE = 1000
    NUM_THREADS = 3
    
    @@results = {}
    @@doneFiles = []
    @@jobQueue = Queue.new
    @@deltaCount = 0
    @@resultLock = Mutex.new
    @@jobqueueLock = Mutex.new
    
    def self.getCsvFilesInDir(path)
        files = []
        Dir.foreach(path) { |f|
            files << f if f.match(/\.csv$/)
        }
        
        return files.sort
    end
    
    def self.resultKey(f1, f2)
        tmp = [ f1, f2 ]
        tmp = tmp.sort
        return tmp[0] + "," + tmp[1]
    end
    
    def self.resultsExist(f1, f2)
        return (@@doneFiles.find_index(resultKey(f1,f2)) != nil) ? true : false
    end
    
    def self.addResultsForEquals(f1, f2)
        res = { "UPDATE" => 0, "INSERT" => 0, "DELETE" => 0 }
        addResults(f1, f2, res)
    end
    
    def self.addResults(f1, f2, res)
        @@resultLock.synchronize {
            @@results[resultKey(f1, f2)] = res
        }
    end
    
    def self.addFromExistingResults(f1, f2, key)
        @@resultLock.synchronize {
            @@results[resultKey(f1, f2)] = @@results[key]
        }
    end
    
    def self.printResult(f1, f2, res)
        return f1 + "," + f2 + "," + 
            res["INSERT"].to_s + "," + 
            res["DELETE"].to_s + "," + 
            res["UPDATE"].to_s
    end
    
    #---------------------------------------------------------------------------
    # GETTER / SETTER
    
    def self.results()
        return @@results
    end
    
    def self.addJob(job)
        @@jobQueue << job
    end
    
    def self.raiseDeltaCount
        @@deltaCount += 1
    end
    
    def self.deltaCount()
        return @@deltaCount
    end
    
    def self.addDoneFiles(key)
        @@doneFiles << key
    end

end

#===============================================================================
# PROCESSING

files = Delta::getCsvFilesInDir(Delta::WORKDIR)
if files.length == 0
    puts "No input files. Exiting."
    exit Delta::EXIT_NO_INPUT_FILES
end

#-------------------------------------------------------------------------------
# Run delta computation for all files

resultsToCopy = []
for f1 in files
    for f2 in files
        
        if Delta::resultsExist(f1, f2)
            puts "Results for " + Delta::resultKey(f1,f2) + " already done."
            resultsToCopy << [ f1, f2 ]
            next
        end
        
        if f1 == f2
            Delta::addResultsForEquals(f1, f2)
        else
            Delta::addJob([f1, f2])
            Delta::raiseDeltaCount
            Delta::addResultsForEquals(f1, f2) # Remove !!
        end
        
        Delta::addDoneFiles(Delta::resultKey(f1, f2))
        
    end
end

# Run threads
threads = []
for i in 0..Delta::NUM_THREADS
    #threads[i] = Thread.new { Delta::startCalculator }
end
for t in threads
    t.join
end

# Copy missing results
for p in resultsToCopy
    Delta::addFromExistingResults(p[0], p[1], Delta::resultKey(p[0],p[1]))
end

# Print results
results = Delta::results
for f1 in files
    for f2 in files
        puts Delta::printResult(f1, f2, results[Delta::resultKey(f1, f2)])
    end
end

puts "Calculated " + Delta::deltaCount.to_s + " out of " + (files.length**2).to_s + 
    " deltas."
